sig-network/api-extensions-best-practices.md

```diff
@@ -0,0 +1,141 @@
+# API Extensions - Best Practices
+
+It is sometimes desireable to build new networking features as [CRDs] rather
+than [in-tree]. Reasons may include (but aren't limited to):
+
+* A desire to be free of lockstep with the [KEP] process
+* Experimentation and highly iterative development
+* The feature is useful to multiple parties, but is niche
+* the API doesn't or can't have a default implementation in core
+
+Whatever the case may be, we have multiple SIG Network sub-projects that have
+taken this approach for various reasons:
+
+* [Gateway API]
+* [Network Policy]
+* [Multi Network]
+
+This document intends to share guidelines and best practices for how to create
+and manage these kinds of "official, but add-on and optional" features, which
+throughout this document we'll refer to as **Official CRDs**.
+
+> **Note**: **This is not a standard**, but **SIG Network projects should try
+> and follow this guidance** unless they have some good reasons not to, as the
+> guidance here comes from the experience of those who have walked this road
+> before. This guidance might be applicable to some other SIGs, but as of the
+> time of writing none except SIG Network have reviewed this.
+
+> **Note**: This is **not a comprehensive guide**.
+
+[CRDs]:https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/
+[in-tree]:https://github.com/kubernetes/api
+[KEP]:https://github.com/kubernetes/enhancements/blob/master/keps/README.md
+[Gateway API]:https://github.com/kubernetes-sigs/gateway-api
+[Network Policy]:https://github.com/kubernetes-sigs/network-policy-api
+[Multi Network]:https://github.com/kubernetes-sigs/multi-network
+
+## Objectives
+
+At a high level, SIG Network sub-projects that deploy as [CRDs] will need to
+deal with (at least) the following dimensions:
+
+* Enhancement Proposals
+* Developing CRDs
+* Delivering CRDs
+* CRD Lifecycle Management
+* Conformance Tests & Reports
+* Documentation
+* Community & Evangelism
+
+In the following section we'll talk about each of these in-depth, including
+some lessons learned by existing or historical sub-projects.
+
+[CRDs]:https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/
+
+### Enhancement Proposals
+
+Projects which produce optional [Custom Resource Defintion (CRD)][CRD]-based
+APIs should provide a project-specific enhancement proposal system. You can see
+examples of using project-specific enhancement proposals with some of the
+current sub-projects:
+
+* [Gateway Enhancement Proposals (GEPs)][geps]
+* [Network Policy Enhancement Proposals (NPEPs)][npeps]
```

Comment:
Cluster API, while under a different sig, also has similar problems I think. They use CAEPs. If you want to expand the scope of this beyond network at some point, might be good to include as an example

Reply:
Yes, making this more broad is definitely on the table. Making SIG Network the starting point and scope limit for the first iteration is mostly about wanting to get something out there quicker at the networking scope to help our upcoming multi-networking efforts, and potentially other projects like KNI.